app/models/pfmp.rb

```diff
@@ -43,6 +47,17 @@ class Pfmp < ApplicationRecord
 
   delegate :wage, to: :mef
 
+  def self.perfect
+    joins(:schooling, student: :rib)
+      .merge(Schooling.student)
+      .merge(Schooling.with_attributive_decisions)
+      .merge(Student.lost.invert_where)
+      .merge(Student.lives_in_france)
+      .merge(Rib.unique)
+      .merge(Pfmp.this_year)
+      .where.not(amount: 0) # FIXME
+  end
+
```

Comment:
C'est un scope ça

Reply:
Et si on pouvait avoir un meilleur terme que "perfect" aussi, plus informatif peut être "candidate_first_batch_asp", je sais pas (avec un comment pour expliquer le truc)

Reply:
c'est absolument un scope, je l'ai juste mis dans une méthode de classe parce que c'est une vraie entreprise d'assemblage de scopes, en opposition à des scopes qui filtrent beaucoup plus simplement la donnée donc je préfère faire cette disctinction.

> When your scope logic gets complicated, a class method feels like the right place to put it.

https://www.justinweiss.com/articles/should-you-use-scopes-or-class-methods/ 

Pour ce qui est du nom, le truc c'est que `first_batch` c'est un peu trop chronologique, en vrai ce scope représente l'ensemble des critères nécessaires pour que l'ASP accepte le dossier. On a `.asp_ready` dans `Student`, on pourrait avoir `Pfmp.asp_ready` ou `Pfmp.asp_complete` ? Votons

Reply:
Fais-en quand même un scope, on peut vouloir le chain avec d'autres morceaux de requêtes ActiveRecord

Reply:
c'est déjà un scope, déjà chainable (`Pfmp.perfect.order(:end_date)`), mais vu la complexité des 7 scopes qu'on enchaînent dedans je préfère que ce soit une méthode de classe pour mieux signaler ça ?

Reply:
En gros le state :asp_ready c'est le vrai state :validated après exploration de la réalité du process d'acceptation ? Si c'est le cas il est capital de le modéliser.

Reply:
En faire une méthode de classe ne signale rien de spécial. Ou alors je ne connais pas cette convention.
Je continue de penser qu'il est préférable d'en faire un scope. Je ne vois pas trop le rapport avec le fait qu'il soit si gros.

Reply:
https://courses.bigbinaryacademy.com/learn-rubyonrails/class-methods-and-scopes/

> Which one to prefer?

> It's all about consistency. Scopes are usually used when the logic is very small or, for simple where/order clauses.

> Class methods are used when it involves a bit more complexity, and when we need a finer grain of control over the execution of queries.

C'est une convention que j'ai déjà rencontré personnellement, comme dit précédemment il y a 7 scopes enchaînés donc je préfère une méthode de classe pour la distinguer des filtres de base que font les autres scopes. C'est sensiblement la même chose au final (scope vs class methods), mais je préfère avec la distinction.

Reply:
> C'est sensiblement la même chose au final
Nope.
Je ne peux pas chain des méthodes de classes quand je fait mes requêtes active record

Reply:
https://github.com/rails/rails/blob/main/activerecord/lib/active_record/scoping/named.rb#L75-L82